### ISSUE TITLE
request to save individual channels

### ISSUE THREAD

**Issue by hhrhhr** [dbdexter-dev/meteor_decode#2]
1) It would be nice to save not a flattened “color” image, but each channel separately for further processing.
2) As I understand it, a JPG stream is coming from the satellite. Maybe you should add an option to save it in its original form without transcoding to png? In theory, just add the jpg header.

**Comment by dbdexter-dev** [comment]
1. That's a nice feature indeed. Commit 7303a37 implements the "-S, --split" flag, which does just that.
2. While the satellite does encode the image following the JPEG format, it splits the image into 112x8 blocks before compressing it, so it may encode each of them with different compression ratios. Outputting a .jpg would therefore require more than just adding a JPEG header to the raw image. I could do it, but the result would be the same as converting the output .png to .jpg with an external tool.

**Comment by hhrhhr** [comment]
1. Wow! But still I want even more ;)
   1. You save one gray channel in RGB, which in my case gives a total of 20 MB PNG files. However, if you save each channel as a gray PNG, then you get 11 MB, almost 2 times less.
Here is an example patch for this: https://gist.github.com/hhrhhr/0abb39b2fba297285bb16f291398d077/revisions?diff=unified
   2. Most people probably just need a color picture, but sometimes it may take both, RGB and gray, not just RGB or gray. That is, two separate options are required.
On the other hand, if you include channel numbers in the name of the RGB file (for example, _LRPT_YYYY_mm_dd-HH_MM-(68-65-64).png_), then we can simply extract each channel separately, that is, the above patch is not needed at all.
On the third hand, soon more than 3 channels can be activated on M2-2, and it is advisable to have them separately.
2. I mean that PNG is not really suitable for storing photos. After converting JPG -> PNG -> JPG the picture usually does not get better.
And of course, this is a question of dependencies ;) To build under Windows for the decoder needs libpng and zlib. By the way, it would be great if they were optional, something like define _USE_PNG=1_ in Makefile.

**Comment by dbdexter-dev** [comment]
1.i: You're absolutely right, I forgot about grayscale PNGs. Will fix.
1.ii: I don't think M2-2 can cram 4 or more channels worth of images in 72k/80k symbols per second. I can probably work on that though, and make it so that you can save as many channels as you want.

2. What I'm trying to say is that every JPG block sent is a valid JPG by itself, but joining them together can result in an invalid JPG, because the encoding parameters can change from one block to the next.  I would still have to decode the blocks into raw pixels first and then re-encode them into a JPG, and the result would be the same as going JPG -> PNG -> JPG. It would also mean exchanging libpng for a JPG encoding library, so the number of dependencies would be the same anyway.

**Comment by hhrhhr** [comment]
Ok, with jpg and png everything is clear. Anyway PNG is better than just uncompressed images ;)

Regarding the separate preservation of channels.
M2_LRPT_Decoder has the ability to save VCDU and APID70. I still have no idea what it is, but the M22TelemetryUnpaker utility (Windows only too :( ) has already appeared, which from * .70 pulls out the coordinates and even the position of the satellite in the form of roll/pitch/yaw, maybe there is other data too.
I tried to save the channel 70 by analogy with 64-69 (not like PNG, naturally, I just wanted to save the raw data), but I didn’t get anything like the result of the Windows utility. can you add the ability to export this channel?

**Comment by hhrhhr** [comment]
structure of APID70:
```
struct packet {
    // big endian!
    // 64 bytes
    char hex1[22];  // unknown data
    uint8 h;        // hour
    uint8 m;        // min
    uint8 s;        // sec
    uint32 ms;      // ms
    char hex2[35];  // unknown data

    // 64 bytes
    uint32 time1;   // number of seconds since the beginning of t_year
    uint16 t_year;
    uint16 t_days;  // number of dats since the beginning of t_year
    uint32 time2;   //

    uint32 time3;
    float qw;       // satelite position (quaternion)
    float qx;
    float qy;
    float qz;

    uint32 time4;
    double X;       // satelite coordinates in ECEF model
    double Y;       // x axis from ellipsoid center to 0°N 0°W, y — to 0°N 90°W, z — to 90°N
    double Z;       // altitude = (X^2 + Y^2 + Z^2)^0.5

    uint32 zero;    // end of packet
}
```

example: https://gist.github.com/hhrhhr/21e31678c10e4354c1503c91efbc1586

**Comment by the-gobleo-tinkerer** [comment]
@hhrhhr APID 70 dump has now been implemented.

**Comment by dbdexter-dev** [comment]
Yeah, like @shuyuan-liu mentioned, I got around to implementing writing APID70 data to a file. Tried it on a couple of M2-2 samples I had lying around, and the telemetry unpacker seems to generate a valid TLE from the dumps. Not of much use now that M2-2 is not transmitting LRPT anymore, but once M2-3 is launched it should work.
